content/microservices/kafka.md

```diff
@@ -0,0 +1,283 @@
+
+```typescript
+@@filename(hero.controller)
+@Controller()
+export class HeroService {
+  @MessagePattern('hero.kill.dragon')
+  killDragon(message: any): any {
+    const realm = 'Nest';
+    const heroId = message.value.heroId;
+    const dragonId = message.value.dragonId;
+
+    const items = [
+      { id: 1, name: 'Mythical Sword' },
+      { id: 2, name: 'Key to Dungeon' },
+    ];
+
+    return {
+      headers: {
+        realm
+      },
+      key: heroId,
+      value: items
+    }
+  }
+}
+```
+
+Additionally, messages passed in this format can also contain custom headers set in the `headers` hash property. Header hash property values must be either of type `string` or type `Buffer`.
+
+```typescript
+@@filename(hero.controller)
+@Controller()
+export class HeroService {
+  @MessagePattern('hero.kill.dragon')
+  killDragon(message: any): any {
+    const realm = 'Nest';
+    const heroId = message.value.heroId;
+    const dragonId = message.value.dragonId;
+
+    const items = [
+      { id: 1, name: 'Mythical Sword' },
+      { id: 2, name: 'Key to Dungeon' },
+    ];
+
+    return {
+      headers: {
+        kafka_nestRealm: realm
+      },
+      key: heroId,
+      value: items
+    }
+  }
+}
+```
+
+#### Naming conventions
+
+The Kafka microservice components append a description of their respective role onto the `client.clientId` and `consumer.groupId` options to prevent collisions between Nest microservice client and server components. By default the `ClientKafka` components append `-client` and the `ServerKafka` components append `-server` to both of these options.
+
+```typescript
+@@filename(main)
+const app = await NestFactory.createMicroservice(ApplicationModule, {
+  transport: Transport.KAFKA,
+  options: {
+    client: {
+      clientId: 'hero', // hero-server
+      brokers: ['localhost:9092'],
+    },
+    consumer: {
+      groupId: 'hero-consumer' // hero-consumer-server
+    },
+  }
+});
+```
+
+```typescript
+@@filename(hero.controller)
+@Client({
+  transport: Transport.KAFKA,
+  options: {
+    client: {
+      clientId: 'hero', // hero-client
+      brokers: ['localhost:9092'],
+    },
+    consumer: {
+      groupId: 'hero-consumer' // hero-consumer-client
+    }
+  }
+})
+client: ClientKafka;
+```
+
+> info **Hint** Kafka client and consumer naming conventions can be customized by extending `ClientKafka` and `KafkaServer` in your own custom provider and overriding the constructor.
+
+Since the Kafka microservice message pattern utilizes two topics for the request and reply channels, a reply pattern should be derived from the request topic. By default, the name of the reply topic is the composite of the request topic name with `.reply` appended.
```

Comment:
I have not been able to listen to messages/event when only using this client.

Do I need to always create the microservice  (NestFactory.createMicroservice()) to listen to messages?

Can I have different consumers per application or creating the microservice will make it to always use the `Kafka Server`?

Reply:
I think my problem was that I was trying to  listen to events from service and not a controller

Reply:
@kamilmysliwiec I have a similar question - the consuming of messages does not seem to  be working unless I pass the same config  to both the server and the client. What is the role of the kafka server created by NestFactory.createMicroservice()?